torch/nested/_internal/cached_tensor.py

```diff
@@ -0,0 +1,125 @@
+from contextlib import contextmanager
+from typing import *  # noqa: F403
+
+import torch
+
+
+class CachedTensor(torch.Tensor):
+    metadata: Dict[str, Optional[torch.Tensor]]
+    source_field: str
+
+    # Tensor subclass wrapping a dict of tensors, whose shape, dtype, device, etc.
+    # is determined by a "source" tensor in the dict (specified by the user
+    # during construction).
+    #
+    # This tensor does not support any operations on it by default.
+    # To leverage the extra metadata, you must register an op to perform the special logic
+    # you want via register_cached_tensor_func.
+    #
+    # When used this way (1) it is a convenient way to keep around metadata
+    # related to a tensor, without having to laboriously thread those extra metadata
+    # around, e.g. through aten signatures. (2) allows one to trigger custom
+    # __torch_dispatch__ logic to construct tensor subclasses (which can be tricky to do
+    # otherwise because the subclass constructor op itself usually does not take the
+    # subclass itself as input!). See NestedTensor for an example.
+
+    @staticmethod
+    @torch._disable_dynamo  # type: ignore[misc]
+    def __new__(
+        cls,
+        metadata: Dict[str, Optional[torch.Tensor]],
+        source_field: str,
+    ) -> "CachedTensor":
+        source = metadata.get(source_field)
+        assert source is not None
+        shape = source.shape
+        kwargs = {}
+        kwargs["strides"] = source.stride()
+        kwargs["storage_offset"] = source.storage_offset()  # type: ignore[assignment]
+        kwargs["device"] = source.device  # type: ignore[assignment]
+        kwargs["layout"] = source.layout  # type: ignore[assignment]
+        kwargs["requires_grad"] = source.requires_grad  # type: ignore[assignment]
+        kwargs["dtype"] = source.dtype  # type: ignore[assignment]
+        out = torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+        out.metadata = metadata
+        out.source_field = source_field
+        return out
+
+    def __repr__(self) -> str:  # type: ignore[override]
+        return f"CachedTensor({repr(self.metadata[self.source_field])})"
+
+    def __getattr__(self, name: str) -> Optional[torch.Tensor]:
+        if name not in self.metadata:
+            raise AttributeError(
+                f"{type(self).__name__} object has no attribute '{name}'"
+            )
+        return self.metadata[name]
+
+    def __tensor_flatten__(self) -> Tuple[List[str], Dict[str, Any]]:
+        ctx = {
+            "source_field": self.source_field,
+            "all_fields": list(self.metadata.keys()),
+        }
+        return [k for k, v in self.metadata.items() if v is not None], ctx
+
+    @staticmethod
+    def __tensor_unflatten__(
+        inner_tensors: Dict, meta: Dict, outer_size: Any, outer_stride: Any
+    ) -> "CachedTensor":
+        inner_tensors = {k: inner_tensors.get(k) for k in meta["all_fields"]}
```

Comment:
aren't `inner_tensors.keys()` and `meta["all_fields"]` identical / redundant? Or is it possible for them to be different?

Reply:
Inner tensors keys are incomplete because they only contain the k's that are non-None. (I'm not sure if we can have inner tensors that are None.)